### DISCUSSION TITLE
Dubbo Java Roadmap

### DISCUSSION THREAD

**Discussion by chickenlj** [apache/dubbo#13880] [General]
For Dubbo3, there are currently three major active versions: `3.1`, `3.2`, and `3.3` in the community. Among them, `3.2` is the recommended stable version while `3.3` is currently in the `beta` stage which will replace `3.2` as the latest stable version in the coming months.

As of version `3.3.0-beta.1`, the core features we have planned for Dubbo3 will be covered. So from this version on, Dubbo3 will enter the long-term stability maintenance phase, which means that we will not introduce any big features for any new Dubbo3 versions in the future, and the community will focus on iteration and stability assurance of existing features. Based on this, we have made the Roadmap as listed below:

> We will of course continue to explore new features in Dubbo4 and keep the community updated.

* 3.3.x iteration
* Lightweight RPC API
* Remoting and protocol layer refactor
* Project and user experience improve
* Documentation construction
* Extension repository apache/dubbo-spi-extensions
* Continuous benchmark
* Kubernetes best practice
* Native Image

## 中文版本

对于 Dubbo3 而言，当前社区维护有 3.1、3.2、3.3 三个大的版本，其中 3.2 是当前推荐的稳定生产版本，3.3 正处于 beta 阶段，在接下来几个月将取代 3.2 成为最新稳定版本。 

截止 3.3.0-beta.1 版本，整个 Dubbo3 规划的核心功能均已开发完成。从这个版本开始，Dubbo3 将进入长期稳定维护阶段，这意味着我们将不会再为任何的 Dubbo3 新版本引入新的功能规划，社区将专注在已有功能的迭代与稳定性保证上。基于这样的基调，我们制定了下文中的 Roadmap。

> 关于面向未来的新功能，我们将在 Dubbo4 中持续探索，欢迎持续关注社区进展。

### 3.3.0-beta 系列迭代
* 参与人员：刘军
* Issue 链接：

### 轻量化
* 参与人员：江河清
* Issue 链接：

- [ ] 轻量化方案设计与实现

### remoting 与 protocol 层重构
* 参与人员：陈有为、梁贝宁、孙彩荣
* Issue 链接：

- [ ] rest 与 triple 协议整合
- [ ] rest 与 rest-metadata 合并
- [ ] rest、triple、remoting-http12、remoting-http
- [ ] remoting-zookeeper 迁移（暂时放configcenter？）

### 项目规范化 & 影响体验的关键问题
* 参与人员：华钟明、刘军
* Issue 链接：

- [ ] Github完善：readme、contributor guide；issue & pull request 相关规范；分支梳理；主干demo整理，保持简洁最佳实践
- [ ] 联合intellij脚手架推广与完善；start.aliyun 中加入dubbo组件
- [ ] 版本与升级指南，各个版本到 3.x 的升级文档（更多内容，站在用户视角，小白用户用dubbo3到底有哪些障碍）
- [ ] FAQ 部分完善
- [x] zookeeper、curator5 版本以及传递依赖问题，尤其是curator5
- [x] spring application.yml 自动配置提示
- [ ] 日志相关问题：文档、使用最佳实践、默认行为、配置、依赖冲突等
- [ ] 配置属性整理，要大而全的，包含yml/properties/xml属性格式的
- [ ] 梳理模块整合、打包诉求
- [x] 清理module：monitor、container、一些工具module
- [x] dubbo-filter、dubbo-native 移动到 plugin
- [ ] distribution 发布包整理
- [ ] 环境变量的配置模式：基本规律、注意事项（单个registry、多个registries、单个registry+id等）
- [x] 当前支持的所有非规范 jvm properties，有表格列出来

### 文档升级
* 参与人员：刘军、全员
* Issue 链接：

- [ ] 文档升级：这部分会拆解几个任务分头执行（涵盖多语言库）；开发者一眼能看到怎么开发、如何加依赖；能一下子看到当前有哪些dubbo版本、jdk/sdk 依赖兼容情况

### extension 库
* 参与人员：宋小生
* Issue 链接：

- [ ] Extension 库和 Dubbo3 适配的总体情况梳理
- [ ] Extenion 库的运作机制、验收机制等
- [ ] 当前所有的spi接口，哪些是重点面向用户的，哪些是内部解耦的，哪些是没必要的；adaptive动态代码的问题

ref: [dubbo extended tasks（扩展库规划）](https://github.com/apache/dubbo-spi-extensions/issues/232)

### benchmark 机制
* 参与人员：王晓彬
* Issue 链接：

- [ ] benchmark 机制建设，要看到一个能用来推广的大图，用户一看到就能眼前一亮的那种，最好是能持续更新（涵盖rpc和地址发现）

### kubernetes 最佳实践
* 参与人员：丛国庆
* Issue 链接：

- [ ] kubernetes 最佳实践，各种生命周期接口等。
- [ ] 后续接 dubboctl 部分
- [ ] xds 未来如何发展？

### 一些最佳实践
* 参与人员：刘军
* Issue 链接：

- [ ] springboot如何迁移到dubbo
- [ ] rest相关的使用场景
- [ ] 流量规则的排查机制

### Native
* 参与人员：华钟明
* Issue 链接：

- [x] 静态化示例整理
- [x] docker 打包
- [ ] 完善native的集成测试
- [ ] sae 部署，出差异化文档和最佳实践推广
- [ ] JVM CRaC 沟通验证
- [ ] 清晰的定义native产品支持边界，给一个支持组件、最佳实践建议的文档
- [ ] 用户案例

**Comment by wxbty** [comment]
Github完善 是否加个历史issue治理，现在open的900多

**Comment by songxiaosheng** [comment]
性能这里可以深度挖掘下，每天跑一次把遗留的和性能问题提前挖掘出来，一起优化下

**Comment by songxiaosheng** [comment]
dubbo admin 与dubbo兼容低版本这一块是升级用户的通用的一个痛点可以看看能不能专门来挖掘一下做一些适配

**Comment by shuangpai** [comment]
虚拟线程也要马上兼容一下，JDK21马上发射了。

**Comment by chickenlj** [comment]
Triple 协议对比 Dubbo 协议的使用限制、不兼容列表

**Comment by olamy** [comment]
Hi there,
I had a look to upgrade Jetty version you are using.
I noticed you are in 9.4.x which is end of support life, 10/11 will be end of support as well.
Upgrade to Jetty 12 means jdk 17 requirement.
Do you have any plan to upgrade jdk requirement soon?

**Comment by chickenlj** [comment]
> Hi there, I had a look to upgrade Jetty version you are using. I noticed you are in 9.4.x which is end of support life, 10/11 will be end of support as well. Upgrade to Jetty 12 means jdk 17 requirement. Do you have any plan to upgrade jdk requirement soon?

The majority of Dubbo users are now using jdk 8. So I think we might do that on a new branch like `3.4`, but the latest branch we are working on is 3.3, so it might take several months.

**Comment by wdaike** [comment]
3.3.0预计几月份可以正式发布，期待virtual thread功能
